compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -550,8 +552,53 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                     match project::poly_project_and_unify_type(self, &project_obligation) {
                         Ok(Ok(Some(mut subobligations))) => {
-                            self.add_depth(subobligations.iter_mut(), obligation.recursion_depth);
-                            self.evaluate_predicates_recursively(previous_stack, subobligations)
+                            'compute_res: {
+                                // If we've previously marked this projection as 'complete', thne
+                                // use the final cached result (either `EvaluatedToOk` or
+                                // `EvaluatedToOkModuloRegions`), and skip re-evaluating the
+                                // sub-obligations.
+                                if let Some(key) =
+                                    ProjectionCacheKey::from_poly_projection_predicate(self, data)
+                                {
+                                    if let Some(cached_res) = self
+                                        .infcx
+                                        .inner
+                                        .borrow_mut()
+                                        .projection_cache()
+                                        .is_complete(key)
+                                    {
+                                        break 'compute_res Ok(cached_res);
+                                    }
+                                }
+
+                                self.add_depth(
+                                    subobligations.iter_mut(),
+                                    obligation.recursion_depth,
+                                );
+                                let res = self.evaluate_predicates_recursively(
+                                    previous_stack,
+                                    subobligations,
+                                );
+                                // If the result is something that we can cache, then mark this
+                                // entry as 'complete'. This will allow us to skip evaluating the
+                                // suboligations at all the next time we evaluate the projection
+                                // predicate.
+                                if res == Ok(EvaluatedToOk) || res == Ok(EvaluatedToOkModuloRegions)
+                                {
+                                    if let Some(key) =
+                                        ProjectionCacheKey::from_poly_projection_predicate(
+                                            self, data,
+                                        )
+                                    {
+                                        self.infcx
+                                            .inner
+                                            .borrow_mut()
+                                            .projection_cache()
+                                            .complete(key, res.unwrap());
```

Comment:
I feel like you should be able to get around this `unwrap` with an if-let and binding...